File: src/lib/model-provider/factory.ts
/**
 * Factory for creating LLM provider instances
 * Supports environment-based provider selection (Groq for dev, Gemini for prod)
 */

import pino from 'pino';
import { GroqProvider } from './groq-provider';
import { GeminiProvider } from './gemini-provider';
import { ModelProviderType } from './types';
import type { BaseModelProvider } from './base-provider';

const logger = pino({ name: 'model-provider-factory' });

// Singleton instances for serverless optimization
let groqInstance: GroqProvider | null = null;
let geminiInstance: GeminiProvider | null = null;

/**
 * Create or retrieve a model provider based on environment configuration
 * Uses singleton pattern for serverless function reuse
 *
 * @param providerType - Optional override for provider type (defaults to LLM_PROVIDER env var)
 * @returns BaseModelProvider instance (Groq or Gemini)
 * @throws Error if provider type is invalid or API key is missing
 */
export function createModelProvider(providerType?: ModelProviderType): BaseModelProvider {
  // Determine provider type from param or env
  const provider =
    providerType ||
    (process.env.LLM_PROVIDER?.toLowerCase() as ModelProviderType) ||
    ModelProviderType.GROQ;

  logger.info({ provider, msg: 'Creating model provider' });

  switch (provider) {
    case ModelProviderType.GROQ: {
      if (!process.env.GROQ_API_KEY) {
        throw new Error(
          'GROQ_API_KEY environment variable is required when LLM_PROVIDER=groq'
        );
      }

      // Return singleton instance if exists
      if (groqInstance) {
        logger.debug({ msg: 'Reusing existing Groq provider instance' });
        return groqInstance;
      }

      logger.info({ msg: 'Initializing new Groq provider instance' });
      groqInstance = new GroqProvider({
        apiKey: process.env.GROQ_API_KEY,
        defaultConfig: {
          model: process.env.GROQ_MODEL || 'llama-3.3-70b-versatile',
          temperature: process.env.LLM_TEMPERATURE
            ? parseFloat(process.env.LLM_TEMPERATURE)
            : 0.7,
          maxTokens: process.env.LLM_MAX_TOKENS
            ? parseInt(process.env.LLM_MAX_TOKENS, 10)
            : 2048,
        },
      });

      return groqInstance;
    }

    case ModelProviderType.GEMINI: {
      if (!process.env.GEMINI_API_KEY) {
        throw new Error(
          'GEMINI_API_KEY environment variable is required when LLM_PROVIDER=gemini'
        );
      }

      // Return singleton instance if exists
      if (geminiInstance) {
        logger.debug({ msg: 'Reusing existing Gemini provider instance' });
        return geminiInstance;
      }

      logger.info({ msg: 'Initializing new Gemini provider instance' });
      geminiInstance = new GeminiProvider({
        apiKey: process.env.GEMINI_API_KEY,
        defaultConfig: {
          model: process.env.GEMINI_MODEL || 'gemini-2.0-flash-exp',
          temperature: process.env.LLM_TEMPERATURE
            ? parseFloat(process.env.LLM_TEMPERATURE)
            : 0.7,
          maxTokens: process.env.LLM_MAX_TOKENS
            ? parseInt(process.env.LLM_MAX_TOKENS, 10)
            : 2048,
        },
      });

      return geminiInstance;
    }

    default:
      throw new Error(
        `Invalid LLM_PROVIDER: ${provider}. Must be 'groq' or 'gemini'`
      );
  }
}

/**
 * Reset singleton instances (useful for testing)
 */
export function resetProviderInstances(): void {
  groqInstance = null;
  geminiInstance = null;
  logger.info({ msg: 'Provider instances reset' });
}

File: src/backend/app/services/llm/factory.py
"""
Factory for creating LLM provider instances.

Provides convenient methods to create providers from configuration.
"""

import logging
from typing import Optional

from app.core.config import settings

from .base import LLMProvider
from .gemini import GeminiLLMProvider

logger = logging.getLogger(__name__)


def create_llm_provider(
    provider_name: Optional[str] = None,
    api_key: Optional[str] = None,
    model: Optional[str] = None,
    temperature: Optional[float] = None,
    max_tokens: Optional[int] = None,
) -> LLMProvider:
    """
    Create an LLM provider instance.

    Args:
        provider_name: Name of the provider ("groq" or "gemini", default: from settings)
        api_key: API key (if None, uses from settings)
        model: Model name (if None, uses from settings)
        temperature: Temperature (if None, uses from settings)
        max_tokens: Max tokens (if None, uses from settings)

    Returns:
        LLMProvider instance

    Raises:
        ValueError: If provider is unknown or API key is missing
    """
    provider = (provider_name or settings.default_llm_provider).lower()
    temp = temperature if temperature is not None else settings.llm_temperature
    max_tok = max_tokens if max_tokens is not None else settings.llm_max_tokens

    if provider == "groq":
        # Groq is deprecated - fallback to Gemini
        logger.warning(f"Groq provider no longer supported - using Gemini instead")
        return create_llm_provider(provider_name=settings.default_llm_provider, api_key=api_key, model=model, temperature=temperature, max_tokens=max_tokens)

    elif provider == "gemini":
        key = api_key or settings.gemini_api_key
        if not key:
            raise ValueError("Gemini API key is required (GEMINI_API_KEY env var)")

        # Use default from config as single source of truth
        model_name = model or settings.default_llm_model
        logger.info(f"Creating Gemini LLM provider with model={model_name}")
        return GeminiLLMProvider(
            api_key=key,
            model=model_name,
            temperature=temp,
            max_tokens=max_tok,
        )

    else:
        raise ValueError(f"Unknown LLM provider: {provider}")


def create_llm_provider_from_env() -> LLMProvider:
    """
    Create an LLM provider from environment configuration.

    Returns:
        LLMProvider instance configured from settings

    Raises:
        ValueError: If configuration is invalid
    """
    return create_llm_provider()
